lib/Dialect/Arc/Transforms/LegalizeStateUpdate.cpp

```diff
@@ -126,128 +87,153 @@ struct AccessAnalysis : public DataFlowAnalysis {
+        if (blockNode->argAccesses.contains({calleeArg, AccessType::Read}))
+          add(*callOpNode, {callArg, AccessType::Read});
+        if (blockNode->argAccesses.contains({calleeArg, AccessType::Write}))
+          add(*callOpNode, {callArg, AccessType::Write});
+      }
+    }
   }
-  propagateIfChanged(argAccesses, result);
 
-  // Update the block's access list.
-  auto *blockAccesses = getOrCreate<AccessState>(block);
-  result = blockAccesses->join(innerAccesses);
-  propagateIfChanged(blockAccesses, result);
+  return success();
+}
+
+OpAccesses *AccessAnalysis::lookup(Operation *op) {
+  return opAccesses.lookup(op);
 }
 
-void AccessAnalysis::visitOperation(Operation *op) {
-  auto result = ChangeResult::NoChange;
-  auto *accesses = getOrCreate<AccessState>(op);
-
-  TypeSwitch<Operation *>(op)
-      .Case<StateReadOp>([&](auto readOp) {
-        result |= accesses->add(readOp.getState(), AccessState::Read);
-      })
-      .Case<StateWriteOp>([&](auto writeOp) {
-        result |= accesses->add(writeOp.getState(), AccessState::Write);
-      })
-      .Case<CallableOpInterface>([&](auto callableOp) {
-        if (auto *region = callableOp.getCallableRegion()) {
-          auto argResult = ChangeResult::NoChange;
-          auto *argAccesses = getOrCreate<ArgumentAccessState>(op);
-          for (auto &block : *region) {
-            argResult |= argAccesses->join(
-                *getOrCreateFor<ArgumentAccessState>(op, &block));
-          }
-          propagateIfChanged(argAccesses, argResult);
-        }
-      })
-      .Case<CallOpInterface>([&](auto callOp) {
-        auto calleeOp = dyn_cast_or_null<CallableOpInterface>(
-            callOp.resolveCallable(&symbolTable));
-        if (!calleeOp)
-          return;
-        auto operands = callOp.getArgOperands();
-        const auto *calleeArgAccesses =
-            getOrCreateFor<ArgumentAccessState>(callOp, calleeOp);
-        for (auto access : calleeArgAccesses->accesses) {
-          auto arg =
-              access.getPointer().template dyn_cast_or_null<BlockArgument>();
-          if (arg)
-            result |=
-                accesses->add(operands[arg.getArgNumber()], access.getInt());
-        }
-      });
+BlockAccesses *AccessAnalysis::lookup(Block *block) {
+  return blockAccesses.lookup(block);
+}
 
-  // Don't propagate inner state accesses through models, clock trees, and
-  // passthrough ops.
-  if (!isa<ModelOp, ClockTreeOp, PassThroughOp>(op)) {
-    for (auto &region : op->getRegions()) {
-      for (auto &block : region) {
-        const auto *blockAccesses = getOrCreateFor<AccessState>(op, &block);
-        result |= accesses->join(*blockAccesses);
-      }
+void AccessAnalysis::add(OpAccesses &op, Access access) {
+  // Propagate to the parent block and operation if the access escapes escapes
+  // the block or targets a blcok argument.
+  if (op.accesses.insert(access).second) {
+    if (op.parent) {
```

Comment:
Nit: We could avoid a level of nesting here:
```suggestion
  if (op.accesses.insert(access).second && op.parent) {
```